[#7] 회원정보 수정/회원탈퇴

cyj199637 opened the pull request.

cyj199637 committed:
[#7] 회원정보 수정 기능 구현

- 회원정보 수정 기능 구현
- 비밀번호 변경 기능 구현중

cyj199637 committed:
[#7] 회원정보 수정/회원탈퇴 기능 구현

- 비밀번호 변경 기능 구현
- 회원탈퇴 기능 구현

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -71,4 +73,56 @@
         return RESPONSE_OK;
     }
+
+    @PutMapping("/profile/{userId}")
```
이런 곳은 `/users/{userId}/profiles` 이런식으로 표현해보면 어떨까요?
아래에 있는 부분도 이 방식을 따라갈 수 있겠네요~

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -71,4 +73,56 @@
+        User currentUser = (User) httpSession.getAttribute("user");
+        if (currentUser == null) {
+            return RESPONSE_UNAUTHORIZED;
+        } else {
```
위에서 유저가 없으면 리턴하므로 else는 굳이 쓸 필요가 없어보이네요~

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -71,4 +73,56 @@
+
+        User currentUser = (User) httpSession.getAttribute("user");
+        if (currentUser == null) {
+            return RESPONSE_UNAUTHORIZED;
```
로그인이 필요한 API에서 계속 아래의 로직이 반복되고있습니다. 이 반복되는 로직을 지울 수 있지 않을까요?

1. 세션에서 로그인한 유저의 정보를 가져온다.
2. 로그인한 유저의 정보가 없으면 에러를 리턴한다

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -71,4 +73,56 @@
+        } else {
+            try {
+                userService.updateUserPassword(currentUser, userPasswordUpdateParam);
+                httpSession.invalidate();
```
이 라인은 로그아웃을 의미하는 것 같은데요. 이런 로그인과 관련된 로직을 한 곳에서 처리해주는 클래스로 분리할 수 있지 않을까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -71,4 +73,56 @@
+
+                return RESPONSE_OK;
+            } catch (InValidValueException e) {
+                return RESPONSE_CONFLICT;
```
`CONFLICT`라는 코드로 `잘못된 값을 입력했다`는 의미를 전달할 수 있을까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -71,4 +73,56 @@
+        }
+    }
+
+    @DeleteMapping("/{userId}")
```
보통 로그인한 유저는 `userId`를 세션에서 가져오는데 path variable로 userId를 중복으로 가져올 필요가 있을까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/UserServiceImpl.java:
```diff
@@ -63,4 +63,58 @@ public User getLoginUser(UserLoginInfo userLoginInfo) {
+        );
+
+        if (!isValidPassword ||
+            userPasswordUpdateParam.getExistPassword().equals(userPasswordUpdateParam.getNewPassword()) ||
```
보통 이렇게 비교하면 `userPasswordUpdateParam.getExistPassword()`가 null이라면 NPE가 날 수 있기 때문에 `StringUtils.equals`를 사용하는게 좋습니다~

cyj199637 committed:
[#7] 피드백 반영

- 로그인 체크 AOP 적용
- UserController Mapping URL 수정
- UserController updateUserPassword InValidValueException 발생 시 Response Body에 메시지 추가
- UserServiceImpl NPE 방지를 위한 StringUtils 사용

f-lab-dev commented on README.md:
```diff
@@ -1,18 +1,36 @@
+* 사용자
+
+  1. 회원가입 / 탈퇴
+
```
굳이 개행은 안넣어도 될 것 같습니다~

f-lab-dev commented on pom.xml:
```diff
@@ -96,6 +96,27 @@
+        -->
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-aspects</artifactId>
```
우리는 spring boot를 사용하니까 spring boot에 속한 라이브러리를 쓰는게 어떨까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/aspect/CheckLoginAspect.java:
```diff
@@ -0,0 +1,27 @@
+@Component
+@Aspect
+public class CheckLoginAspect {
+    @Before("@annotation(me.liiot.snsserver.annotation.CheckLogin)")
```
이 위에 개행 하나 해주면 깔끔할 것 같습니다~

f-lab-dev commented on src/main/java/me/liiot/snsserver/aspect/CheckLoginAspect.java:
```diff
@@ -0,0 +1,27 @@
+    @Before("@annotation(me.liiot.snsserver.annotation.CheckLogin)")
+    public void checkLogin() throws HttpClientErrorException {
+        HttpServletRequest servletRequest =
+                ((ServletRequestAttributes) RequestContextHolder.getRequestAttributes()).getRequest();
```
`RequestContextHolder.getRequestAttributes()`가 null일 수 있을까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/aspect/CheckLoginAspect.java:
```diff
@@ -0,0 +1,27 @@
+                ((ServletRequestAttributes) RequestContextHolder.getRequestAttributes()).getRequest();
+        HttpSession httpSession = servletRequest.getSession();
+
+        if (httpSession.getAttribute("user") == null) {
```
`"user"`라는 세션키는 상수로 사용하는게 좋을 것 같습니다~ 이전에 리뷰드렸던 세션관련 클래스 분리와 묶어서 생각해보셔도 되겠네요

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -74,55 +75,48 @@
 
-    @PutMapping("/profile/{userId}/password")
-    public ResponseEntity<Void> updateUserPassword(UserPasswordUpdateParam userPasswordUpdateParam,
+    @PutMapping("/{userId}/profiles/passwords")
```
1. password는 `profiles`에 속한 1개이기때문에 단수로 해주셔도 됩니다~
2. 자신의 비밀번호를 바꾸는것일텐데 path variable로 userId를 꼭 받아야할까요? 세션에서 가져오면 될텐데요~

cyj199637 committed:
[#7] 피드백 반영

- UserController Mapping URL 수정
- SessionUtil로 세션 메소드 분리
- SessionKey에 있는 세션 키 상수 SessionUtil로 이동

f-lab-dev commented on pom.xml:
```diff
@@ -107,17 +107,18 @@
-            <artifactId>spring-aspects</artifactId>
-            <version>5.2.7.RELEASE</version>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-aop</artifactId>
```
`spring-aspects`와 `spring-boot-starter-aop`의 차이는 무엇일까요?

f-lab-dev commented on pom.xml:
```diff
@@ -107,17 +107,18 @@
-            <version>5.2.7.RELEASE</version>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-aop</artifactId>
+            <version>2.3.1.RELEASE</version>
```
버전을 꼭 명시해줘야할까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/aspect/CheckLoginAspect.java:
```diff
@@ -14,13 +15,14 @@
         HttpSession httpSession = servletRequest.getSession();
 
-        if (httpSession.getAttribute("user") == null) {
+        if (httpSession.getAttribute(SessionUtil.USER) == null) {
```
읽을 때 `SessionKeys.USER` 이렇게 쓰는게 어떤 역할인지 더 쉽게 알 수 있지 않을까요??

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -64,55 +64,55 @@
     }
 
-    @PutMapping("/{userId}/profiles")
+    @PutMapping("/myaccount")
```
`/my-account`로 단어를 구분해주면 좋겠습니다~

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -64,55 +64,55 @@
         }
 
-        httpSession.setAttribute(SessionKey.USER, user);
+        SessionUtil.setUser(httpSession, user);
```
`setUser`라는 메소드명보다 역할로 명시해주는게 낫지 않을까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/util/SessionUtil.java:
```diff
@@ -0,0 +1,23 @@
+
+import javax.servlet.http.HttpSession;
+
+public class SessionUtil {
```
이쪽은 역할을 세션에 집중하는게 아니라 비즈니스적으로 풀어주면 좋을 것 같은데요~ 
`LoginService` 인터페이스로 추상화시켜주고, 이 로그인은 세션방식 로그인이므로 `SessionLoginService` 이런식으로 확장해주면 더 객체지향적이지 않을까요?

cyj199637 committed:
[#7] 피드백 반영

- 로그인과 관련된 로직 LoginService로 분리

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/SessionLoginService.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    private HttpSession getSession() {
+        HttpServletRequest servletRequest =
```
`HttpServletRequest` 혹은 `HttpSession`은 스프링에서 주입받을 수 있습니다~

cyj199637 committed:
[#7] Custom Parameter Annotation 적용

- 세션에서 로그인한 사용자 객체를 받아오기 위한 Annotation 생성
- Annotation이 적용된 파라미터에 대해 적용할 ArgumentResolver 생성

f-lab-dev commented on src/main/java/me/liiot/snsserver/annotation/GetCurrentUser.java:
```diff
@@ -0,0 +1,11 @@
+
+@Target({ElementType.PARAMETER})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface GetCurrentUser {
```
이름이 `GetCurrentUser`가 적절할까요? 어노테이션은 동작을 나타내는 것이 아닙니다~
`CurrentUser` 정도가 적절해보입니다

f-lab-dev reviewed (commented):
AOP를 로우하게 사용하지 않고 딱 스프링의 정석대로 하셨네요. 좋습니다~👍

cyj199637 committed:
[#7] 피드백 반영

- CurrentUser로 어노테이션 이름 변경
